### PR TITLE
preparation for glutin without memory leak

### PR DESCRIPTION
**Pull request by bogdad**
Hi @aeickhoff !  

There was a fix committed to the master of glutin, fixing memory leak for @autoreleasepool content.

screenshot 2017-08-26 17 51 25

this is the link to glutin pr:
https://github.com/tomaka/glutin/pull/925

Please consider merging this pr which updates glutin, glium and imgui to the latest, so it would be easier to update to latest glutin and glium and imgui when they are released.

The difference is in glium - GlutinFacade got renamed to Display, in winit the events structure changed, and you have to provide event loop to glutin facade.

Just to stress - this does not fix memory leak yet, waiting for new glutin release

Best wishes,
Vladimir.

### CODE REVIEW

**Comment by aeickhoff** [comment]
You rock @bogdad ! Thank you for making me aware of this problem with glutin and getting everything up to date, something that I wanted to do anyways for some time already. I’ll take a detailed look at it in the evening.

File: engine/stagemaster/src/lib.rs [hunk 1]
```diff
@@ -24,25 +26,32 @@ pub mod camera_control;
...
 use std::collections::BTreeMap;
 
 pub struct UserInterface {
-    window: GlutinFacade,
+    events_loop: EventsLoop,
+    interactables: HashMap<ID, (AnyShape, usize)>,
+    ui_inner: UserInterfaceInner,
+}
+
+struct UserInterfaceInner {
```
Comment (aeickhoff) [inline engine/stagemaster/src/lib.rs]: Why did you split it into `UserInterface` and `UserInterfaceInner`? And why are `interactables` not in `UserInterfaceInner`?

File: engine/stagemaster/src/combo.rs [hunk 2]
```diff
@@ -293,7 +293,17 @@ impl ComboListener {
     pub fn update(&mut self, event: &Event) {
         let old_current = self.current;
         let something_changed = match *event {
```
Comment (aeickhoff) [inline engine/stagemaster/src/combo.rs]: Can we match on `event.event` to make the two following match arms simpler?

File: engine/stagemaster/src/lib.rs [hunk 3]
```diff
@@ -156,118 +169,131 @@ pub fn setup(
...
 
         the_ui.on_critical(move |_: &ProcessEvents, ui, world| {
-            let scale = ui.imgui.display_framebuffer_scale();
+            let scale = ui.ui_inner.imgui.display_framebuffer_scale();
 
-            for event in ui.window.poll_events().collect::<Vec<_>>() {
+            let ui_inner = &mut ui.ui_inner;
+
+            ui.events_loop.poll_events(|event| {
                 match event {
```
Comment (aeickhoff) [inline engine/stagemaster/src/lib.rs]: Again, can we match `event.event` to keep the match arms simpler?

**Comment by bogdad** [comment]
Sure, the rustfmt already suggested that, will do in few days.

Regarding the ui_inner - the lambda for processevents wanted to borrow the
whole ui and event loop and failed because event_loop is already mutably
borrowed as we are calling the method on it. Splitting event loop out of
actual ui data helped. I could try fixing it differently, not very sure
though.

File: engine/stagemaster/src/lib.rs (hunk 1 above)
Comment (bogdad) [reply engine/stagemaster/src/lib.rs]: fixed, thanks!

File: engine/stagemaster/src/combo.rs (hunk 2 above)
Comment (bogdad) [reply engine/stagemaster/src/combo.rs]: fixed, thanks!

File: engine/stagemaster/src/lib.rs (hunk 1 above)
Comment (bogdad) [reply engine/stagemaster/src/lib.rs]: it occured to me that it was just possible to get all events from the event loop and there is no need to have a 'large' lambda, just as it was previously.

File: engine/stagemaster/src/lib.rs (hunk 3 above)
Comment (bogdad) [reply engine/stagemaster/src/lib.rs]: fixed, thanks!

File: engine/stagemaster/src/lib.rs (hunk 1 above)
Comment (bogdad) [reply engine/stagemaster/src/lib.rs]: so there is no need for in the first place splitting!

**Comment by bogdad** [comment]
Hi @aeickhoff ,
Fixed the match on event.event, reverted the split of UserInterface, please consider merging!

**Comment by aeickhoff** [comment]
Thanks, I'll have a look at it later today!

**Comment by aeickhoff** [comment]
FYI: now merging this into my latest work. Unfortunately an ugly merge (my fault), I'll take care of it. Thanks for implementing my requested changes!

**Comment by aeickhoff** [comment]
Merged 🎉

**Comment by bogdad** [comment]
Cool!)

**Pull request merged**